### PR TITLE
Generate Painless Factory for Creating Script Instances

### PR DESCRIPTION
**Pull request by jdconrad**
Uses bytecode generation in Painless to create a factory that generates script instances based on a specific context.

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScriptEngine.java [hunk 1]
```diff
@@ -130,10 +139,78 @@ public boolean needsScores() {
...
+                }
+            });
+
+            compile(contextsToCompilers.get(context), loader, scriptName, scriptSource, params);
+
+            return generateFactory(loader, context);
+        }
+    }
+
+    private <T> T generateFactory(Loader loader, ScriptContext<T> context) {
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScriptEngine.java]: Can you please add a javadoc here giving a high level overview of what is generated (ie ctor with which params, members, etc)?

**Review by rjernst — APPROVED** [review]
LGTM

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScriptEngine.java (hunk 1 above)
Comment (jdconrad) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScriptEngine.java]: Done.

**Comment by jdconrad** [comment]
@rjernst Thanks for the review.

**Pull request merged**